crypto/x509/x509_vfy.c

```diff
@@ -1784,31 +1785,54 @@ static int internal_verify(X509_STORE_CTX *ctx)
      */
     while (n >= 0) {
         /*
+         * For each iteration of this loop:
+         * n is the subject depth
+         * xs is the subject cert, for which the signature is to be checked
+         * xi is the supposed issuer cert containing the public key to use
+         * Initially xs == xi if the last cert in the chain is self-issued.
+         *
          * Skip signature check for self-signed certificates unless explicitly
          * asked for because it does not add any security and just wastes time.
-         * If the issuer's public key is not available or its key usage does
-         * not support issuing the subject cert, report the issuer certificate
-         * and its depth (rather than the depth of the subject).
          */
-        if (xs != xi || (ctx->param->flags & X509_V_FLAG_CHECK_SS_SIGNATURE)) {
+        if (xs != xi || ((ctx->param->flags & X509_V_FLAG_CHECK_SS_SIGNATURE)
+                         && (xi->ex_flags & EXFLAG_SS) != 0)) {
```

Comment:
I am not sure about the EXFLAG_SS check here. This will be set only for V3 certs with AKID / SUBJECT key id. I'd still prefer if we force the signature verification with X509_V_FLAG_CHECK_SS_SIGNATURE even for V1 certs. If I have just self signed end-entity V1 certificate and I set X509_V_FLAG_CHECK_SS_SIGNATURE, I'd expect it to be verified.

Reply:
Good thought to make sure that is works also for V1 certs, but I don't see a problem at this point:
For such certs there is no AKID and then for self-issued certs the EXFLAG_SS is always set.
This is directly visible by looking at the definition of `X509_check_akid()`.

Reply:
Good, I did not investigate the code in depth. Sorry for the noise.